### PR TITLE
ENH: Adds interfaces for MRtrix utils shconv and sh2amp

### PR DESCRIPTION
**Pull request by tclose**
## Summary

This PR adds the predicted signal output option of MRtrix's dwi2tensor two interfaces to mrtrix3.utils: SHConv and SH2Amp, which wrap the MRtrix3 (>v3.0.0) commands `shconv` and `sh2amp` respectively. Together these can be useful for calculating the residual signal after fitting tensor or ODF models.

## List of changes proposed in this PR (pull-request)

* Adds predicted_signal output to FitTensor interface
* Adds mrtrix3.SHConv interface
* Adds mrtrix3.SH2Amp interface

## Acknowledgment

- [x] \(Mandatory\) I acknowledge that this contribution will be available under the Apache 2 license.

### CODE REVIEW

File: nipype/interfaces/mrtrix3/reconst.py [hunk 1]
```diff
@@ -187,6 +203,9 @@ def _list_outputs(self):
             outputs["gm_odf"] = op.abspath(self.inputs.gm_odf)
         if self.inputs.csf_odf != Undefined:
             outputs["csf_odf"] = op.abspath(self.inputs.csf_odf)
+        if self.inputs.predicted_signal != Undefined:
```
Comment (effigies) [inline nipype/interfaces/mrtrix3/reconst.py:206]: I assume that you're being consistent with the rest of this function. If you feel the urge, you can fix up the others, but I think we should do this the standard way.

```suggestion
        if isdefined(self.inputs.predicted_signal):
```

File: nipype/interfaces/mrtrix3/utils.py [hunk 2]
```diff
@@ -765,3 +765,132 @@ class MRResize(MRTrix3Base):
...
+    checking residuals of ODF estimates.
+
+
+    Example
+    -------
+
+    >>> import nipype.interfaces.mrtrix3 as mrt
+    >>> sh = mrt.SHConv()
+    >>> sh.inputs.in_file = 'odf.mif'
+    >>> sh.inputs.response = 'response.txt'
```
Comment (effigies) [inline nipype/interfaces/mrtrix3/utils.py:816]: These files should exist in https://github.com/nipy/nipype/tree/master/nipype/testing/data. If there isn't already one with a name that works for this tool, feel free to create an empty file with that name.

File: nipype/interfaces/mrtrix3/utils.py [hunk 3]
```diff
@@ -765,3 +765,132 @@ class MRResize(MRTrix3Base):
...
+    >>> sh.cmdline                               # doctest: +ELLIPSIS
```
Comment (effigies) [inline nipype/interfaces/mrtrix3/utils.py:817]: No `...` appears in the output line.

```suggestion
    >>> sh.cmdline
```

File: nipype/interfaces/mrtrix3/utils.py [hunk 4]
```diff
@@ -765,3 +765,132 @@ class MRResize(MRTrix3Base):
...
+    checking residuals of ODF estimates.
+
+
+    Example
+    -------
+
+    >>> import nipype.interfaces.mrtrix3 as mrt
+    >>> sh = mrt.SH2Amp()
+    >>> sh.inputs.in_file = 'sh.mif'
+    >>> sh.inputs.directions = 'grads.txt'
```
Comment (effigies) [inline nipype/interfaces/mrtrix3/utils.py:883]: Again, these should exist as dummy files.

File: nipype/interfaces/mrtrix3/utils.py [hunk 5]
```diff
@@ -765,3 +765,132 @@ class MRResize(MRTrix3Base):
...
+    >>> sh.cmdline                               # doctest: +ELLIPSIS
```
Comment (effigies) [inline nipype/interfaces/mrtrix3/utils.py:884]:
```suggestion
    >>> sh.cmdline
```

**Review by effigies — COMMENTED** [review]
Overall looks good. Some small comments.

**Comment by effigies** [comment]
Also can you run `black` on the files you've touched and `make specs`?

File: nipype/interfaces/mrtrix3/reconst.py (hunk 1 above)
Comment (tclose) [reply nipype/interfaces/mrtrix3/reconst.py]: On this interface, I remembered that 'predicted_signal' is only a valid option when using the 'msmt_csd' algorithm. Is there a way to specify this in the input spec?

**Comment by tclose** [comment]
Ok, I think I have made all the changes you requested now

**Review by effigies — COMMENTED** [review]
Regarding `msmt_csd`, please see #3176 and #3004.

I *think* what we want to do is leave `EstimateFOD` alone and improve `ConstrainedSphericalDeconvolution`, since that one is capable of doing something other than `msmt_csd`. This would encourage people to move to the more "correct" interface to get the full feature set.

Some small suggestions, but otherwise this is looking good.

**Comment by effigies** [comment]
Hey @tclose just a reminder that there are unresolved suggestions here.

**Comment by tclose** [comment]
Sorry, I think I was on leave when they came through and they slipped of the radar.

File: nipype/interfaces/mrtrix3/utils.py [hunk 6]
```diff
@@ -765,3 +765,127 @@ class MRResize(MRTrix3Base):
...
+        mandatory=True,
+        argstr="%s",
+        position=-2,
+        desc=("The response function"),
+    )
+    out_file = File(
+        name_template="%s_shconv.mif",
+        name_source=["in_file"],
+        argstr="%s",
+        mandatory=True,
```
Comment (effigies) [inline nipype/interfaces/mrtrix3/utils.py:790]:
```suggestion
```

File: nipype/interfaces/mrtrix3/utils.py [hunk 7]
```diff
@@ -765,3 +765,127 @@ class MRResize(MRTrix3Base):
...
+        desc=(
+            "The gradient directions along which to sample the spherical "
+            "harmonics MRtrix format"
+        ),
+    )
+    out_file = File(
+        name_template="%s_amp.mif",
+        name_source=["in_file"],
+        argstr="%s",
+        mandatory=True,
```
Comment (effigies) [inline nipype/interfaces/mrtrix3/utils.py:852]:
```suggestion
```

File: nipype/interfaces/mrtrix3/utils.py [hunk 8]
```diff
@@ -765,3 +765,127 @@ class MRResize(MRTrix3Base):
...
+
+    Example
+    -------
+
+    >>> import nipype.interfaces.mrtrix3 as mrt
+    >>> sh = mrt.SHConv()
+    >>> sh.inputs.in_file = 'csd.mif'
+    >>> sh.inputs.response = 'response.txt'
+    >>> sh.cmdline
+    'shconv csd.mif response.txt sh.mif'
```
Comment (effigies) [inline nipype/interfaces/mrtrix3/utils.py:815]: I think this should be:

```suggestion
    'shconv csd.mif response.txt csd_shconv.mif'
```

File: nipype/interfaces/mrtrix3/utils.py [hunk 9]
```diff
@@ -765,3 +765,127 @@ class MRResize(MRTrix3Base):
...
+
+    Example
+    -------
+
+    >>> import nipype.interfaces.mrtrix3 as mrt
+    >>> sh = mrt.SH2Amp()
+    >>> sh.inputs.in_file = 'sh.mif'
+    >>> sh.inputs.directions = 'grads.txt'
+    >>> sh.cmdline
+    'sh2amp sh.mif grads.txt amp.mif'
```
Comment (effigies) [inline nipype/interfaces/mrtrix3/utils.py:880]:
```suggestion
    'sh2amp sh.mif grads.txt sh_amp.mif'
```

**Review by effigies — COMMENTED** [review]
Just about there, @tclose. Can you merge `master`, as well, to get any bug fixes from there?

**Comment by effigies** [comment]
I'm going to try to make a release in the next couple days. If you want to try to finish this off, I'd love to include it!

**Comment by tclose** [comment]
Sorry to miss your release, I'm just getting back into this now. Hopefully it passes the checks now (I'm off to bed:).

On a very tangential note, I have created a new Pydra task repository for Dcm2niix at tclose/pydra-dcm2niix, which is nearly ready to go I think. Would it be good to migrate it into the nipype org, and if so, which is the best way to do this?

**Comment by effigies** [comment]
Oh, hey Tom. We didn't actually release due to some failing checks, so I'll try to get this in.

Re pydra-dcm2niix, would you mind opening an issue on https://github.com/nipype/pydra-tasks-template/issues? We don't really have a process for transitioning projects into nipype control, so this would be a good one to work through manually. At some point, it would be nice to have something like a `conda-forge/feedstocks` setup. (cc @satra I think this is what we discussed?)

**Comment by tclose** [comment]
Ok great, so it is all good to go now?

**Comment by effigies** [comment]
Looks like it! In it goes.

**Pull request merged**